File: prolog/day17_part2_2017.pl
:- initialization(main).

main :-
    read_file_to_string('input.txt', String, []),
    atom_number(String, Steps),
    find_value_after_zero(Steps, ValueAfterZero),
    format('~w', [ValueAfterZero]).

find_value_after_zero(Steps, ValueAfterZero) :-
    find_value_after_zero(Steps, 0, 0, 0, ValueAfterZero).

find_value_after_zero(_, CurrentPos, I, ValueAfterZero, ValueAfterZero) :-
    I >= 50000000, !.

find_value_after_zero(Steps, CurrentPos, I, Acc, ValueAfterZero) :-
    I1 is I + 1,
    CurrentPos1 is (CurrentPos + Steps) mod I1,
    (CurrentPos1 = 0 -> ValueAfterZero1 = I1; ValueAfterZero1 = Acc),
    CurrentPos2 is CurrentPos1 + 1,
    find_value_after_zero(Steps, CurrentPos2, I1, ValueAfterZero1, ValueAfterZero).

File: erlang/day17_part2_2017.erl
-module(task).
-export([call/0]).

call() ->
    {ok, Data} = file:read_file("input.txt"),
    Steps = list_to_integer(binary_to_list(Data)),
    ValueAfterZero = find_value_after_zero(Steps, 50000000, 0, 0),
    io:format("~p~n", [ValueAfterZero]).

find_value_after_zero(Steps, N, CurrentPos, ValueAfterZero) ->
    find_value_after_zero(Steps, N, CurrentPos, ValueAfterZero, 1).

find_value_after_zero(_, 0, _, ValueAfterZero, _) -> ValueAfterZero;
find_value_after_zero(Steps, N, CurrentPos, ValueAfterZero, I) ->
    NewPos = (CurrentPos + Steps) rem I,
    NewValueAfterZero = if NewPos == 0 -> I; true -> ValueAfterZero end,
    find_value_after_zero(Steps, N - 1, NewPos + 1, NewValueAfterZero, I + 1).